02-microarray/pathway-analysis_microarray_07_gsea.Rmd

```diff
@@ -0,0 +1,443 @@
+
+The function we will use to map from Ensembl gene IDs to Entrez gene IDs is called `mapIds()`.
+
+Let's create a column in the `dge_df` data frame that shows the mapped Entrez IDs along with the information for the respective Ensembl IDs.
+
+```{r}
+# We would like a data frame we can join to the differential expression stats
+dge_df <- dge_df %>%
+  # Create a new column 'entrez_id' that contains the Entrez IDs returned by
+  # mapIds()
+  dplyr::mutate(entrez_id = mapIds(org.Dr.eg.db,
+    keys = dge_df$Gene,
+    column = "ENTREZID",
+    keytype = "ENSEMBL",
+    multiVals = "first"
+  )) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id))
+```
+
+This message is letting us know that sometimes Ensembl gene identifiers will map to multiple Entrez IDs.
+In this case, it's also possible that a gene symbol will map to multiple Ensembl IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of the new `dge_df`. 
+
+```{r}
+head(dge_df)
+```
+
+## Perform gene set enrichment analysis (GSEA)
+
+**DRAFT**
+Gene set enrichment testing using `clusterProfiler` is based on... [@clusterProfiler-book].
+
+### Determine our pre-ranked genes list
+
+The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+This is step 1 -- gene-level statistics.
+
+```{r}
+lfc_vector <- dge_df$logFC
+names(lfc_vector) <- dge_df$entrez_id
+lfc_vector <- sort(lfc_vector, decreasing = TRUE)
+```
+
+```{r}
+# Look at first entries of the log2 fold change vector
+head(lfc_vector)
+```
+
+### Run GSEA using the `GSEA()` function
+
+**DRAFT**
+The enrichment score (ES) for a pathway, a pathway-level statistic, is calculated using our gene-level statistics.
+Genes are ranked from most highly positive to most highly negative and weighting them according to their gene-level statistic.
+A running score is calculated by starting with the most highly ranked genes and increasing the score when a gene is in the pathway and decreasing the score when a gene is not in the pathway.
+The ES is the maximum deviation from zero.
+Significance is assessed by generating a null distribution by sampling random gene sets of the same size and an FDR (false discovery rate) value is calculated to account for multiple hypothesis testing.
+([Subramanian 2005](https://doi.org/10.1073/pnas.0506580102); [Korotkevich 2019](https://doi.org/10.1101/060012)).
+
+**REVIEW**
+We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).
+
+```{r}
+gsea_results <- GSEA(
```

Comment:
Good point, I'll look into this.

Reply:
Ah. Okay, I wasn't sure what you meant by `arbitrarily small P-values`; it looks like `fgseaMultilevel()` does resampling with Monte Carlo simulations whereas  `fgseaSimple()` does not (just runs it once). So `fgseaMultilevel()` re-runs the analysis with a bunch of different subsets to get more validated p values. 

We can probably just run the `Simple` one? What do you think? But also if we do use `fgseaMultilevel()` we'll need to look into setting the seed.

Reply:
This will likely be elaborated in PR 2 of 2, but it seems that `fgseaMultilevel()` is the recommended approach to use and the default for `clusterProfiler::GSEA()` when the `nPerm` argument is not provided (info gathered partly from docs and partly from the warning message below):

<img width="788" alt="Screen Shot 2020-10-29 at 1 09 40 PM" src="https://user-images.githubusercontent.com/43576623/97742535-cb2e9900-1aba-11eb-9ba3-b7d047df3d99.png">

Also, when attempting to run `fgseaSimple()`, 0 results were returned (could be an error in my implementation however).

Reply:
Great! I was wondering if you saw any recommendations. Let's use what they recommend then!